synapse/lib/config.py

```diff
@@ -1,8 +1,126 @@
+        self.types = {}
+        self._initTypes()
+
+    def _initTypes(self):
+        # These are super primitive and offer little safety
+        # for example, nearly anything can be cast into str()
+        self.addType('int', int)
+        self.addType('str', str)
+        self.addType('bool', bool)
+        self.addType('dict', dict)
+        self.addType('list', list)
+        self.addType('float', float)
+
+    def addType(self, name: str, func: typing.Callable):
+        self.types[name] = func
+
+    def norm(self, name: str, valu: typing.Any) -> typing.Any:
+        func = self.types.get(name)
+        if func is None:
+            raise s_exc.NoSuchName(mesg='no norm func for name', name=name)
+        ret = func(valu)
+        return ret
+
+class Config2:
+    def __init__(self, confdefs, prefix='SYN_'):
+        self.normer = ConfTypes()
+        self.prefix = prefix
+        self.confdefs = confdefs
+        self.conf = {}
+        self.norms = {}
+        self.yaml_loads = ('dict',
+                           'list',
+                           )
+
+        for key, info in self.confdefs:
+            self.norms[key] = info.get('type').lower()
+            defval = info.get('defval', s_common.novalu)
+            if defval is s_common.novalu:
+                continue
+            self.conf[key] = copy.deepcopy(defval)
+
+    def get(self, name, default=None):
+        return self.conf.get(name, default)
+
+    # Container magic methods
+    def __setitem__(self, key, valu):
+        self._set(key, valu)
+
+    def __getitem__(self, item):
+        return self.conf[item]
+
+    # dict methods
+    def copy(self):
+        return copy.deepcopy(self)
+
+    def items(self):
+        return self.conf.items()
+
+    def values(self):
+        return self.conf.values()
+
+    def keys(self):
+        return self.conf.keys()
+
+    async def set(self, name, valu):
+        self._set(name, valu)
+
+    def _set(self, name, valu):
+        norm = self.norms.get(name)
+        if norm is None:
+            logger.warning(f'Config key: [{name}] has no norm function.')
+            self.conf[name] = valu
+            return
+            # XXX THIS only fails in the test suite because the
+            # CompatTest.test_compat_cellauth hsa some archaic config
+            # data in it.  However, warning and moving along does allow
+            # someone to cheat by shoveling config data in that isn't
+            # neccesarily used. Tradeoffs...
+            # raise s_exc.NoSuchName(name=name)
+        self.conf[name] = self.normer.norm(norm, valu)
+
+    async def loadConfDict(self, conf: typing.Dict):
+        # use a copy of the input dict, since the values may be
+        # modified at runtime by a object.
+        conf = conf.copy()
+
+        for name, valu in conf.items():
+            await self.set(name, valu)
+
+    async def loadConfYaml(self, *path):
+        fp = s_common.genpath(*path)
+        if os.path.isfile(fp):
```

Comment:
including this conditional check and silently doing nothing seems a bit dangerous.  Feels a bit more like the isfile check should be carried out by the caller prior to calling this